cpp/lib/ResultSetArrow.cpp

```diff
@@ -0,0 +1,750 @@
+    {
+        SF_DB_TYPE currSnowflakeType = m_metadata[m_currColumnIdx].type;
+        std::string currName(m_metadata[m_currColumnIdx].name);
+        int64 currScale = m_metadata[m_currColumnIdx].scale;
+        fields.push_back(createArrowField(
+            currSnowflakeType,
+            currName,
+            currScale));
+        m_currColumnIdx++;
+    }
+
+    schema = arrow::schema(fields);
+
+    // Reset column index for future use.
+    m_currColumnIdx = 0;
+
+    return schema;
+}
+
+std::vector<std::shared_ptr<arrow::Array>>
+Snowflake::Client::ResultSetArrow::initColumns(cJSON * data)
+{
+    std::vector<std::shared_ptr<arrow::Array>> columns;
+    std::vector<arrow::ArrayBuilder *> builders = initBuilders(data);
+
+    for (arrow::ArrayBuilder * currBuilder : builders)
+    {
+        std::shared_ptr<arrow::Array> currColumn;
+
+        if (!currBuilder->Finish(&currColumn).ok())
+        {
+            // Error!
+        }
+
+        columns.push_back(currColumn);
+    }
+
+    return columns;
+}
+
+std::vector<arrow::ArrayBuilder *>
+Snowflake::Client::ResultSetArrow::initBuilders(cJSON * data)
+{
+    // Iterate over the column metadata and initialize array builders for each column.
+    std::vector<arrow::ArrayBuilder *> builders;
+    while (m_currColumnIdx < m_totalColumnCount)
+    {
+        SF_DB_TYPE currSnowflakeType = m_metadata[m_currColumnIdx].type;
+        int64 scale = m_metadata[m_currColumnIdx].scale;
+        builders.push_back(createArrowArrayBuilder(currSnowflakeType, scale));
+        m_currColumnIdx++;
+    }
+
+    // Reset column index for future use.
+    m_currColumnIdx = 0;
+
+    // Populate builders with partial results already retrieved in first chunk.
+    appendChunkToBuilders(builders, data);
+
+    // Populate builders with any remaining chunks until all rows are consumed.
+    do
+    {
+        appendChunkToBuilders(
+            builders,
+            m_chunkDownloader->queue[m_currChunkIdx].chunk);
+    }
+    while (advance());
+
+    return builders;
+}
+
+std::shared_ptr<arrow::Field>
+Snowflake::Client::ResultSetArrow::createArrowField(
+    SF_DB_TYPE snowflakeType,
+    std::string name,
+    int64 scale
+)
+{
+    std::shared_ptr<arrow::DataType> arrowType =
+        convertTypeFromSnowflakeToArrow(snowflakeType, scale);
+
+    return arrow::field(name, arrowType);
+}
+
+arrow::ArrayBuilder *
+Snowflake::Client::ResultSetArrow::createArrowArrayBuilder(
+    SF_DB_TYPE snowflakeType,
+    int64 scale
+)
+{
+    switch (snowflakeType)
+    {
+        case SF_DB_TYPE_BINARY:
+        {
+            arrow::StringBuilder * builder =
+                new arrow::StringBuilder(arrow::default_memory_pool());
+            // Since each column should contain exactly as many elements as there are
+            // rows in the result set, resize the array builder for performance.
+            builder->Resize(m_totalRowCount);
+            return builder;
```

Comment:
@SimbaGithub Instead of allocating new memory pool for string builder. We can pass the column from the arrow chunk as is. 
std::shared_ptr<arrow::Array> columnArray = currentBatch->column(i);

Reply:
Sorry for the late comment.
Please resolve if this has been sufficiently addressed. Otherwise, let me know if there are further issues.